// metronom/dist/lib/NodeRedisAdaptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable class-methods-use-this */
const redis_1 = require("redis");
class NodeRedisAdaptor {
    constructor(options) {
        Object.defineProperty(this, "redisClient", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.redisClient = (0, redis_1.createClient)(options);
    }
    hGetAll(redisKey) {
        return this.redisClient.hGetAll(redisKey);
    }
    hSet(redisKey, values) {
        return this.redisClient.hSet(redisKey, values);
    }
    connect() {
        this.redisClient.connect();
    }
    keys(regex) {
        return this.redisClient.keys(regex);
    }
    del(redisKey) {
        return this.redisClient.del(redisKey);
    }
    sendCommand(commands) {
        return this.redisClient.sendCommand(commands);
    }
    echo(message) {
        return this.redisClient.echo(message);
    }
}
exports.default = NodeRedisAdaptor;
